### ISSUE TITLE
[Bug] Import failure: `from ray.rllib.agents.ppo import ppo` fails due to pickling error

### ISSUE THREAD

**Issue by thakkarparth007** [ray-project/ray#20682]
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

RLlib

### What happened + What you expected to happen

**What?**
`from ray.rllib.agents import ppo` should work for v1.8 (right?)

**Logs**
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/scratch/.pyenv/versions/myenv/lib/python3.6/site-packages/ray/cloudpickle/__init__.py in dumps_debug(obj, *args, **kwargs)
     38     try:
---> 39         return dumps(obj, *args, **kwargs)
     40     except (TypeError, PicklingError) as exc:

~/scratch/.pyenv/versions/myenv/lib/python3.6/site-packages/ray/cloudpickle/cloudpickle_fast.py in dumps(obj, protocol, buffer_callback)
     72             )
---> 73             cp.dump(obj)
     74             return file.getvalue()

~/scratch/.pyenv/versions/myenv/lib/python3.6/site-packages/ray/cloudpickle/cloudpickle_fast.py in dump(self, obj)
    619         try:
--> 620             return Pickler.dump(self, obj)
    621         except RuntimeError as e:

~/scratch/.pyenv/versions/myenv/lib/python3.6/site-packages/ray/pickle5_files/pickle5/pickle.py in dump(self, obj)
    484             self.framer.start_framing()
--> 485         self.save(obj)
    486         self.write(STOP)

~/scratch/.pyenv/versions/myenv/lib/python3.6/site-packages/ray/pickle5_files/pickle5/pickle.py in save(self, obj, save_persistent_id)
    600         # Save the reduce() output and finally memoize the object
--> 601         self.save_reduce(obj=obj, *rv)
    602 

~/scratch/.pyenv/versions/myenv/lib/python3.6/site-packages/ray/pickle5_files/pickle5/pickle.py in save_reduce(self, func, args, state, listitems, dictitems, state_setter, obj)
    688         else:
--> 689             save(func)
    690             save(args)

~/scratch/.pyenv/versions/myenv/lib/python3.6/site-packages/ray/pickle5_files/pickle5/pickle.py in save(self, obj, save_persistent_id)
    575                 if reduce is not None:
--> 576                     rv = reduce(self.proto)
    577                 else:

TypeError: can't pickle function objects

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
<ipython-input-1-13d560222efb> in <module>
      1 import ray
      2 from ray import tune
----> 3 from ray.rllib.agents import ppo

~/scratch/.pyenv/versions/myenv/lib/python3.6/site-packages/ray/rllib/__init__.py in <module>
     56 
     57 _setup_logger()
---> 58 _register_all()
     59 
     60 __all__ = [

~/scratch/.pyenv/versions/myenv/lib/python3.6/site-packages/ray/rllib/__init__.py in _register_all()
     33     for key in list(ALGORITHMS.keys()) + list(CONTRIBUTED_ALGORITHMS.keys(
     34     )) + ["__fake", "__sigmoid_fake_data", "__parameter_tuning"]:
---> 35         register_trainable(key, get_trainer_class(key))
     36 
     37     def _see_contrib(name):

~/scratch/.pyenv/versions/myenv/lib/python3.6/site-packages/ray/tune/registry.py in register_trainable(name, trainable, warn)
     74         raise TypeError("Second argument must be convertable to Trainable",
     75                         trainable)
---> 76     _global_registry.register(TRAINABLE_CLASS, name, trainable)
     77 
     78 

~/scratch/.pyenv/versions/myenv/lib/python3.6/site-packages/ray/tune/registry.py in register(self, category, key, value)
    148             raise TuneError("Unknown category {} not among {}".format(
    149                 category, KNOWN_CATEGORIES))
--> 150         self._to_flush[(category, key)] = pickle.dumps_debug(value)
    151         if _internal_kv_initialized():
    152             self.flush_values()

~/scratch/.pyenv/versions/myenv/lib/python3.6/site-packages/ray/cloudpickle/__init__.py in dumps_debug(obj, *args, **kwargs)
     45         else:
     46             msg = _warn_msg(obj, "ray.cloudpickle.dumps", exc)
---> 47             raise type(exc)(msg)

TypeError: ray.cloudpickle.dumps(<class 'ray.rllib.agents.trainer_template.A2C'>) failed.
To check which non-serializable variables are captured in scope, re-run the ray script with 'RAY_PICKLE_VERBOSE_DEBUG=1'.
```

### Versions / Dependencies

Ray: 1.8
Python: 3.6.3
OS: Linux 3.10 (CentOS)

### Reproduction script

`from ray.rllib.agents import ppo`. I'm not sure if this is reproduceable that easily, because it worked on colab. I'm not sure what could be the issue on my system.

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

**Comment by thakkarparth007** [comment]
More info. On running with "RAY_PICKLE_VERBOSE_DEBUG" environment variable set, I got these additional logs:

```
===========================================================================
Checking Serializability of <class 'ray.rllib.agents.trainer_template.A2C'>
===========================================================================
!!! FAIL serialization: can't pickle function objects
    Serializing '__getstate__' <function build_trainer.<locals>.trainer_cls.__getstate__ at 0x2ab8fd6567b8>...
    !!! FAIL serialization: can't pickle function objects
    Detected 1 global variables. Checking serializability...
        Serializing 'Trainer' <class 'ray.rllib.agents.trainer.Trainer'>...
        !!! FAIL serialization: Could not pickle object as excessively deep recursion required.
            Serializing '__getstate__' <function Trainer.__getstate__ at 0x2ab8fd5af1e0>...
            !!! FAIL serialization: can't pickle function objects
            Serializing '_allow_unknown_configs' False...
            !!! FAIL serialization: Could not pickle object as excessively deep recursion required.
    Serializing '_allow_unknown_configs' False...
    !!! FAIL serialization: Could not pickle object as excessively deep recursion required.
        Serializing 'denominator' 1...
        !!! FAIL serialization: Could not pickle object as excessively deep recursion required.
            Serializing 'denominator' 1...
            !!! FAIL serialization: Could not pickle object as excessively deep recursion required.
===========================================================================
Variable: 

	FailTuple(_allow_unknown_configs [obj=False, parent=<class 'ray.rllib.agents.trainer.Trainer'>])
FailTuple(__getstate__ [obj=<function Trainer.__getstate__ at 0x2ab8fd5af1e0>, parent=<class 'ray.rllib.agents.trainer.Trainer'>])
FailTuple(denominator [obj=1, parent=1])

was found to be non-serializable. There may be multiple other undetected variables that were non-serializable. 
Consider either removing the instantiation/imports of these variables or moving the instantiation into the scope of the function/class. 
If you have any suggestions on how to improve this error message, please reach out to the Ray developers on github.com/ray-project/ray/issues/
```

**Comment by thakkarparth007** [comment]
The error is actually being thrown by `cloudpickle_fast.py`, and the docstring of this module talks about some functionality only being available for Python 3.8+. Does Ray 1.8 need Python 3.8+? I'm not sure of that because this import works fine on 3.7.3 on my Windows machine.

**Comment by thakkarparth007** [comment]
The underlying error was `maximum recursion depth exceeded in comparison`.  The system I was running on had a max recursion depth of 3000 by default. I even set it to 30K, but the issue still crops up. In comparison, on my Windows machine, the default recursion depth is just 1000, and it just works.

**Comment by thakkarparth007** [comment]
Upgraded python version to 3.6.9 and it works now. Not closing yet though, because there isn't a clear resolution for this issue. Either the documentation/install requirements must be updated or need clarity on why the issue is happening.

**Comment by smestern** [comment]
Can confirm I am running into the same issue (ray 1.8.0, python 3.7.1, win 10). Tried setting the recursion limit to 30k with no fix. Downgrading to ray 1.4.0 also fixes the issue but its missing some features I would like to use.
Will try a fresh env of python 3.8 to see if that remedies the issue.

**Comment by carlogrisetti** [comment]
@thakkarparth007 @smestern Had the same issue, as soon as I upgraded to Python 3.7.12 (no official installer for windows, but unofficial builds are available at https://github.com/adang1345/PythonWindows/tree/master/3.7.12) and rebuilt the venv (don't forget this if you're using one) it worked like a charm :)
Seems to be something bugged or not yet implemented in 3.7.0 (and 3.7.1 apparently)

**Comment by carlogrisetti** [comment]
Also, the latest 3.6.x versions should have this fixed as well, but can't seem to trace back to where I got that info.

@ericl (tagging you at random, divert this as needed) a warning\error in console when using >=3.6.0 <3.6.x or >=3.7.0 < 3.7.x could be a possible solution. Is there any template already done for this reason within the codebase?
